numpy/lib/_function_base_impl.py

```diff
@@ -4717,18 +4719,24 @@ def _quantile(
     # Virtual because it is a floating point value, not an valid index.
     # The nearest neighbours are used for interpolation
     try:
-        method = _QuantileMethods[method]
+        method_props = _QuantileMethods[method]
     except KeyError:
         raise ValueError(
             f"{method!r} is not a valid method. Use one of: "
             f"{_QuantileMethods.keys()}") from None
-    virtual_indexes = method["get_virtual_index"](values_count, quantiles)
+    virtual_indexes = method_props["get_virtual_index"](values_count,
+                                                        quantiles)
     virtual_indexes = np.asanyarray(virtual_indexes)
 
     supports_nans = (
             np.issubdtype(arr.dtype, np.inexact) or arr.dtype.kind in 'Mm')
 
-    if np.issubdtype(virtual_indexes.dtype, np.integer):
+    int_virtual_indices = np.issubdtype(virtual_indexes.dtype, np.integer)
+    supports_integers = method_props["fix_gamma"] is None
+    supports_integers = supports_integers or \
+                            (method == 'linear' and int_virtual_indices)
+
+    if supports_integers:
```

Comment:
I have to think about it a tiny bit more maybe.  We end up with integers here due to integer input.  I suspect we had discussed this, but I am not sure we got it right:  I wonder if we shouldn't just use `method_props["fix_gamma"] is None`.

That would change the odd exception `method == 'linear'` with 0 or 1 as input.  But I rather prefer the clarity that the result type depends (mainly) on the method and not on some (very special!) values.

(Others may have disagreed though, arguing that for 0 and 1 the answer *is* clearly exact for the linear method.)

I like the `fix_gamma` to `None` refactor, though.  And I am not convinced that a `/ 1` to the virtual index calculation (or so... which would also be plausible) would be a better fix.

Reply:
The type indeed depends on the input values: `np.quantile([0,1,2,3,],[0,1])` results in integer type, while `np.quantile([0,1,2,3,],[0,.1])` is float. 

 The reason for making an exception for `method == 'linear'` is to allow for ` tf_quant = np.quantile(True, False)`. That case is in the unit tests, so I added the special case to keep supporting it. 

It seems very odd to run quantile on boolean types though....

Reply:
Ah, I still agree with my above comment:  linear interpolation should always interpolate to ensure the right result dtype, IMO.  Even if the user passed in 0 or 1 to the quantile.
(Assuming this is what is going on, because I didn't think about it very long now.)

Reply:
@seberg Always interpolating for method 'linear' indeed seems the right thing to do. I do not expect any performance loss in the general case (where the quantiles argument will have floating values and not just `0` and `1`). On the other hand:

* The original behaviour of the linear method was to sometimes take the integer path. This makes no difference for common types like `float64`, but might for other input types
* Even if I modify the code to always do interpolation for method linear, then `np.quantile([1,2,1,2,1,2], [0, 1])` results in  `array([1, 2])` (so integer type), while `np.quantile([1,2,1,2,1,2], [0.5])` results in `array([1.5])` (floating type). So always running the interpolation is no guarantee we end up with the same dtype (even for the same dtype input). For the linear method this could be solved by replacing `1` with `1.` somewhere in the code. But again: that works fine for the normal input types (`float32`, `float64`), but might change behavior for other types.
* Forcing linear to always do the interpolation breaks existing code. In particular `np.quantile([True], [False])` fails because `TypeError: numpy boolean subtract, the - operator is not supported, use ...`. How to deal with this? We could just drop support for any input types that do not support `subtract` (for the `linear` method). Another option would be to decide whether to take the integer path based on the input dtype. So `dtype=boolean`: use the integer path, `dtype=float64`: do not use the input path. The latter option does not seem very elegant though, I would not want to have a big table deciding how to handle different input dtypes.